flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerHARecoveryTest.java

```diff
@@ -243,6 +256,268 @@ public void testJobRecoveryWhenLosingLeadership() throws Exception {
+			Future<Object> jobSubmitted = jobManager.ask(
+					new JobManagerMessages.SubmitJob(jobGraph, ListeningBehaviour.DETACHED),
+					deadline.timeLeft());
+
+			Await.ready(jobSubmitted, deadline.timeLeft());
+
+			// Wait for running
+			Future<Object> jobRunning = jobManager.ask(
+					new TestingJobManagerMessages.NotifyWhenJobStatus(jobGraph.getJobID(), JobStatus.RUNNING),
+					deadline.timeLeft());
+
+			Await.ready(jobRunning, deadline.timeLeft());
+
+			// terminate the job manager
+			jobManagerRef.tell(PoisonPill.getInstance(), ActorRef.noSender());
+
+			Future<Boolean> terminatedFuture = Patterns.gracefulStop(jobManagerRef, deadline.timeLeft());
+			Boolean terminated = Await.result(terminatedFuture, deadline.timeLeft());
+			assertTrue("Failed to stop job manager", terminated);
+
+			// job stays in the submitted job graph store
+			assertTrue(mySubmittedJobGraphStore.contains(jobGraph.getJobID()));
+
+			// start new job manager
+			myLeaderElectionService.reset();
+
+			jobManagerRef = createJobManagerActor(
+					"jobmanager-1",
+					flinkConfiguration,
+					myLeaderElectionService,
+					mySubmittedJobGraphStore,
+					500,
+					timeout,
+					jobRecoveryTimeout,
+					archiveRef);
+
+			jobManager = new AkkaActorGateway(jobManagerRef, newLeaderSessionID);
+
+			Future<Object> isAlive = jobManager.ask(TestingMessages.getAlive(), deadline.timeLeft());
+
+			isLeader = jobManager.ask(
+					TestingJobManagerMessages.getNotifyWhenLeader(),
+					deadline.timeLeft());
+
+			isConnectedToJobManager = tmGateway.ask(
+					new TestingTaskManagerMessages.NotifyWhenRegisteredAtJobManager(jobManagerRef),
+					deadline.timeLeft());
+
+			Await.ready(isAlive, deadline.timeLeft());
+
+			// tell new jobManager that he's the leader
+			myLeaderElectionService.isLeader(newLeaderSessionID);
+			// tell taskManager who's the leader
+			myLeaderRetrievalService.notifyListener(jobManager.path(), newLeaderSessionID);
+
+			Await.ready(isLeader, deadline.timeLeft());
+			Await.ready(isConnectedToJobManager, deadline.timeLeft());
+
+			jobRunning = jobManager.ask(
+					new TestingJobManagerMessages.NotifyWhenJobStatus(jobGraph.getJobID(), JobStatus.RUNNING),
+					deadline.timeLeft());
+
+			// wait that the job is recovered and reaches state RUNNING
+			Await.ready(jobRunning, deadline.timeLeft());
+
+			Future<Object> jobFinished = jobManager.ask(
+					new TestingJobManagerMessages.NotifyWhenJobRemoved(jobGraph.getJobID()),
+					deadline.timeLeft());
+
+			BlockingInvokable.unblock();
+
+			// wait til the job has finished
+			Await.ready(jobFinished, deadline.timeLeft());
+
+			// check that the job has been removed from the submitted job graph store
+			assertFalse(mySubmittedJobGraphStore.contains(jobGraph.getJobID()));
+
+			// Check that the BLOB store files are removed
+			File rootPath = new File(haStoragePath);
+
+			boolean cleanedUpFiles = false;
+			while (deadline.hasTimeLeft()) {
+				if (listFiles(rootPath).isEmpty()) {
```

Comment:
We check that the directory no longer contains files. But we don't check for folders, right? I think that we no longer delete the folders created by the BlobStore. We could maybe check in `BlobStore.cleanUp` whether there are any empty folders which we can delete. Do you think that this could be relevant?